docs/design/sql/05-dml.md

```diff
@@ -0,0 +1,100 @@
+# SQL DML
+
+## Overview
+In SQL, the Data Manipulation Language comprises the SQL-data change statements which modify stored data.
+Typically, we can distinguish three classes of commands:
+- **CREATE** which creates new record(s) (`INSERT`)
+- **UPDATE** which updates record(s) fields (`UPDATE`, `MERGE`)
+- **DELETE** which deletes record(s) (`DELETE`)
+
+## 1. INSERT
+To create new record(s) one of two variations of `INSERT` statement can be used:
+- `INSERT INTO` - will append to the target table. If it’s not possible due to unique constraint violation, it will
+  fail. It’s a safe choice if the target never overwrites (such as Kafka, JMS or a JDBC that use a sequence for the
+  primary key or has no primary key)
+- `INSERT OVERWRITE` - will allow overwrites. It will behave as a sequence of `DELETE` followed by an `INSERT`,
+  performed atomically. Specifically, the fields not listed in the insert statement will not retain their values, but
+  will be assigned `null` or default values. If `INSERT OVERWRITE` is used for a target that doesn’t overwrite by
+  design (like Kafka), it will behave exactly as `INSERT INTO`. No error will be thrown. The delete in the imaginary
+  `DELETE` action will delete no rows.
+
+### 1.1 Serialization
+In addition to primitive SQL types (which can be mapped to Java types and stored directly), two serialization formats
+are supported:
+- POJO (reflective access to public properties and fields) - requires Java class on the classpath and no-arg
+  constructor
+- Portable (supported just for IMDG data structures) - if `InMemoryFormat` is set to `BINARY` it requires just
+  `ClassDefinition` to be registered upfront, if `InMemoryFormat` is set to `OBJECT` it requires also Portable factory
+  and Portable class itself on the classpath
+
+Serialization formats have following precedence (the order in which we check whether given serialization can be
+applied):
+1. primitive SQL types
+2. POJO
+3. Portable
+
+Eventual errors from type mismatch between declared and actual types are deferred to statement execution.
+
+The way the data is serialized is expressed via DDL. Depending on the nature of stored objects that information can be
+encoded either in column names or using `OPTIONS` clause.
+
+#### 1.1.1 Key-Value storage
+The serialization formats of key and value are specified separately. To determine whether given field belongs to key or
+value either Java class or Portable `ClassDefinition` is inspected. In case of a name clash between fields of key and
+value composite objects, the ones from key take precedence and are actually written.
+
+To define a table where both key and value are of primitive SQL types, `__key` & `this` as column names should be used.
+For example:
+```
+CREATE EXTERNAL TABLE name (
+  __key INT,
+  this VARCHAR
+) TYPE "com.hazelcast.LocalPartitionedMap"
```

Comment:
One thing that concerns me is the usage of `CREATE EXTERNAL TABLE` for **internal** cluster objects because the whole command context is misleading:
1. Nothing is really created because the map already exists. Likewise, when we call "DROP" the map is not destroyed
1. The map is internal to the system, not external

What if rethink completely how treat maps? Consider that in future we may want to create and drop distributed objects declaratively, @cangencer has some ideas about it. In this case, we will do something like `CREATE MAP myMap ...` or `DROP MAP myMap ...`. 

Now, what we may want to do with an existing `IMap` is to add some metadata, like these class names. This could be done with the `ALTER` statement. E.g.:
`ALTER MAP myMap KEY_CLASS com.PersonKey VALUE_CLASS com.Person TABLE_NAME my_map_table`

WDYT?